### PR TITLE
Update zlib (#31356)

### PR DESCRIPTION
**Pull request by miklelappo**
To mitigate CVE-2022-37434

### CODE REVIEW

**Comment by miklelappo** [comment]
Can't add label...

**Comment by drfloob** [comment]
@miklelappo Thank you for bringing this to our attention. CC'ing @ZhenLian from the security team. 

 From the CVE description: 

> NOTE: only applications that call inflateGetHeader are affected.

gRPC does not directly call `inflateGetHeader`, but we should update the dependency regardless.

The work for updating the dependency is a bit more involved than updating the submodule alone. See the third_party dependency update instructions here https://github.com/grpc/grpc/tree/master/third_party. A new zlib minor release was put out a few days ago, v1.2.13, we should pin to it.

**Comment by miklelappo** [comment]
@drfloob thanks for the hint.
I updated grpc_deps.bzl and check_submodules.sh in sanity checks..
Do I need (or can I even run it myself or do I have no access to Google could?) the next steps:
image

**Comment by miklelappo** [comment]
@ZhenLian android source tree also contains the old zlib version and October security bulletin has no CVE for zlib included. While grpc doesn't directly use this, other packages potentially can (although cs.android.com doesn't find any references), but maybe some 3-rd parties
https://cs.android.com/android/platform/superproject/+/master:external/zlib/zlib.h;l=40?q=zlib

**Comment by miklelappo** [comment]
step 4: generate_projects.sh passes too now

**Review by drfloob — APPROVED** [review]
Thanks, this looks good to me, pending a second look from the security folks. I took care of updating the mirrors.

**Comment by miklelappo** [comment]
@drfloob thanks for your help! Really appreciate that!

File: bazel/grpc_deps.bzl [hunk 1]
```diff
@@ -217,11 +217,11 @@ def grpc_deps():
         http_archive(
             name = "zlib",
             build_file = "@com_github_grpc_grpc//third_party:zlib.BUILD",
-            sha256 = "ef47b0fbe646d69a2fc5ba012cb278de8e8946a8e9649f83a807cc05559f0eff",
-            strip_prefix = "zlib-21767c654d31d2dccdde4330529775c6c5fd5389",
+            sha256 = "90f43a9c998740e8a0db24b0af0147033db2aaaa99423129abbd76640757cac9",
```
Comment (jtattermusch) [inline bazel/grpc_deps.bzl:220]: FTR, this is upgrading from https://github.com/madler/zlib/releases/tag/v1.2.12 to https://github.com/madler/zlib/releases/tag/v1.2.13

**Comment by AshfaqurIntel** [comment]
Is it required to cherry-pick this to v1.50.x branch for the next release?

**Comment by drfloob** [comment]
@AshfaqurIntel the release managers for the last four releases (1.47 through 1.50)  have been notified, and patch-releases will be available soon.

**Comment by jtattermusch** [comment]
@drfloob  probably a good idea to backport to 1.46.x as well, since this is our LTS branch for C++ and C#.

**Comment by frankiefoo-i** [comment]
> @drfloob probably a good idea to backport to 1.46.x as well, since this is our LTS branch for C++ and C#.

@drfloob Is there any work planned for this in near future?

**Comment by psfoley** [comment]
@drfloob thanks for your (and the grpc team's!) work on this. Any chance these backports (specifically 1.48.x) will make their way to PyPi soon?

**Comment by drfloob** [comment]
Quick update: the package builds for 1.46 through 1.48 need to be fixed, the CI system has drifted a bit since then. We're still working on backporting fixes to get those releases out. CC @jtattermusch @gnossen @yashykt.

**Comment by drfloob** [comment]
Update: the 1.46.6 release is available, build issues have been resolved. I've pinged the release managers again for 1.47 and 1.48 again.

Apologies for the delay, gRPC releases are fairly involved processes.

**Comment by drfloob** [comment]
We now have the zlib library upgraded in patch releases from v1.46 onwards. This job is done.

**Pull request merged**